### DISCUSSION TITLE
Marking a Column as deprecated?

### DISCUSSION THREAD

**Discussion by miketheman** [sqlalchemy/sqlalchemy#9309] [Usage Questions] [answered]
Hello!

I've been looking around for ways to accomplish marking an existing `Column` as deprecated inside a codebase, so that usage of it surfaces a `DeprecatingWarning` when used, to assist in refactoring/removal.

I've tried redeclaring the `Column` as a `@property` function that can return the same value, but predictably that met with some nasty `RecursionError`.

Is there some other method that I'm not finding, or would this be a new feature to add a keyword to the `Column` class (or some superclass)?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I'm guessing this in an orm class?

You can probably do that using [validates](https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#sqlalchemy.orm.validates)

**Reply by miketheman** [reply]
Yes, it's an orm class.

I'm guessing that `validates` would work well for when I'm first creating a record, not when calling the specific Column?

I added this to try it out, and I do indeed get the DeprecationWarning on create, not on access.

```python
    @validates("keywords")
    def keywords_deprecation(self, key, value):
        warnings.warn(
            "The keywords attribute is deprecated.",
            DeprecationWarning,
        )
        return value
```

**Reply by CaselIT** [reply]
right. 

Then I guess you can change it as follows.
From
```py
class Foo(Base):
  id = Column(Integer, primary_key=True)
  col_to_deprecate = Column(Integer)
```
To
```py
class Foo(Base):
  id = Column(Integer, primary_key=True)
  _col_to_deprecate = Column("col_to_deprecate", Integer)

  @property
  def col_to_deprecate(self):
    warnings.warn(...)
    return self._col_to_deprecate

  @col_to_deprecate.setter
  def col_to_deprecate(self, value):
    warnings.warn(...)
    self._col_to_deprecate = value
```

**Reply by miketheman** [reply]
That's exactly what I wasn't thinking of - that I could rename the original column 🤦 

I'll give that a shot, thanks!

**Reply by CaselIT** [reply]
the trick is to pass a name to Column so that the table continue to have the old name for the column. The new name is only on the class, not on the db

**Reply by pkong-ds** [reply]
@CaselIT thanks for the suggestion!

Can you provide an example for Declarative table too?

From

```py
from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, mapped_column

class Foo(Base):
  id: Mapped[int] = mapped_column(Integer, primary_key=True)
  col_to_deprecate: Mapped[int] = mapped_column(Integer)
```

To

```
?
```

**Reply by CaselIT** [reply]
should be same:
```py
class Foo(Base):
  id: Mapped[int] = mapped_column(Integer, primary_key=True)
  _col_to_deprecate: Mapped[int] = mapped_column('col_to_deprecate', Integer)
  @property
  def col_to_deprecate(self):
    warnings.warn(...)
    return self._col_to_deprecate

  @col_to_deprecate.setter
  def col_to_deprecate(self, value):
    warnings.warn(...)
    self._col_to_deprecate = value
```

**Comment by fabiomontefuscolo** [comment]
What about other places in the code using queries with `Foo.col_to_deprecate`? It is not going to work as a mapped attribute and SQLA will raise an exception. Can the `@hybrid_property` solve that?

UPDATE: that is what is working for me so far, using `sqlalchemy.ext.hybrid.hybrid_property`

```python
from sqlalchemy import Integer
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Mapped, mapped_column

class Foo(Base):
  id: Mapped[int] = mapped_column(Integer, primary_key=True)
  _col_to_deprecate: Mapped[int] = mapped_column('col_to_deprecate', Integer)

  @hybrid_property
  def col_to_deprecate(self):
    warnings.warn(...)
    return self._col_to_deprecate

  @col_to_deprecate.setter
  def col_to_deprecate(self, value):
    warnings.warn(...)
    self._col_to_deprecate = value

  @col_to_deprecate.expression
  def col_to_deprecate(cls):
    warnings.warn(...)
    return cls._col_to_deprecate
```

Now you can have something like 

```python
query.filter(Foo.col_to_deprecate == "thanks for reading me")
```

**Reply by CaselIT** [reply]
yes, it should work
